### PR TITLE
test: add superchain erc20 bridge tests

### PR DESCRIPTION
**Pull request by agusduha**
Closes OPT-225

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20Factory.sol [hunk 1]
```diff
@@ -27,8 +27,8 @@ contract OptimismSuperchainERC20Factory is IOptimismERC20Factory, ISemver {
     );
 
     /// @notice Semantic version.
-    /// @custom:semver 1.0.0-beta.1
-    string public constant version = "1.0.0-beta.1";
+    /// @custom:semver 1.0.0-beta.3
+    string public constant version = "1.0.0-beta.3";
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/OptimismSuperchainERC20Factory.sol:31]: Wouldn't be only 1 bump here?
```suggestion
    /// @custom:semver 1.0.0-beta.2
    string public constant version = "1.0.0-beta.2";
```

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        vm.mockCall(_receiver, _calldata, _returned);
+        vm.expectCall(_receiver, _calldata);
+    }
+
+    /// @notice Tests the `sendERC20` function burns the sender tokens, sends the message, and emits the `SendERC20`
+    /// event.
+    function testFuzz_sendERC20_succeeds(address _sender, address _to, uint256 _amount, uint256 _chainId) external {
+        // Ensure `_sender` is not the zero address
+        vm.assume(_sender != ZERO_ADDRESS);
+        vm.assume(_to != ZERO_ADDRESS);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: Not needed
```suggestion
```

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        // Check the total supply and balance of `_sender` after the send were updated correctly
+        assertEq(superchainERC20.totalSupply(), _totalSupplyBefore - _amount);
+        assertEq(superchainERC20.balanceOf(_sender), _senderBalanceBefore - _amount);
+    }
+
+    /// @notice Tests the `relayERC20` function reverts when the caller is not the L2ToL2CrossDomainMessenger.
+    function testFuzz_relayERC20_notMessenger_reverts(address _caller, address _to, uint256 _amount) public {
+        // Ensure the caller is not the messenger
+        vm.assume(_caller != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+        vm.assume(_to != ZERO_ADDRESS);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: Not needed
```suggestion
```

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+        // Expect the revert with `CallerNotL2ToL2CrossDomainMessenger` selector
+        vm.expectRevert(ISuperchainERC20Bridge.CallerNotL2ToL2CrossDomainMessenger.selector);
+
+        // Call the `relayERC20` function with the non-messenger caller
+        vm.prank(_caller);
+        superchainERC20Bridge.relayERC20(address(superchainERC20), _caller, _to, _amount);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: NIT: Here the `address(superchainERC20)` token arg could be fuzzed

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            abi.encodeWithSelector(IL2ToL2CrossDomainMessenger.crossDomainMessageSender.selector),
+            abi.encode(_crossDomainMessageSender)
+        );
+
+        // Expect the revert with `InvalidCrossDomainSender` selector
+        vm.expectRevert(ISuperchainERC20Bridge.InvalidCrossDomainSender.selector);
+
+        // Call the `relayERC20` function with the sender caller
+        vm.prank(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+        superchainERC20Bridge.relayERC20(address(superchainERC20), _crossDomainMessageSender, _to, _amount);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: NIT: Here the `address(superchainERC20)` token arg could be fuzzed

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /// @notice Tests the `relayERC20` mints the proper amount and emits the `RelayERC20` event.
+    function testFuzz_relayERC20_succeeds(address _from, address _to, uint256 _amount, uint256 _source) public {
+        vm.assume(_from != ZERO_ADDRESS);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: Not needed
```suggestion
```

File: packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol (hunk 4 above)
Comment (agusduha) [reply packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: hmm what do u mean? it is a `OptimismSuperchainERC20` deployed at the setup

Comment (agusduha) [reply packages/contracts-bedrock/test/L2/SuperchainERC20Bridge.t.sol]: oh i see that u mean is not required to be the same

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20Factory.sol (hunk 1 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/OptimismSuperchainERC20Factory.sol:31]: weird semver diff script error make me do this

**Review by 0xDiscotech — APPROVED** [review]

**Pull request merged**